security/sensorhub-security-oauth/src/main/java/org/sensorhub/impl/security/oauth/OAuthAuthenticator.java

```diff
@@ -258,32 +312,27 @@ public Authentication validateRequest(ServletRequest req, ServletResponse resp,
 
                         OAuthJSONAccessTokenResponse oAuthResponse = oAuthClient.accessToken(authRequest, HttpMethod.POST);
                         accessToken = oAuthResponse.getAccessToken();
-
-                    } catch (OAuthProblemException e) {
-
-                        if (!mandatory)
-                            return Authentication.UNAUTHENTICATED;
-
-                        log.error("Error received from token endpoint: {} - {}", e.getError(), e.getDescription());
-                        response.sendError(HttpServletResponse.SC_FORBIDDEN);
-                        return Authentication.SEND_FAILURE;
-
-                    } catch (OAuthSystemException e) {
-
-                        log.error("Error while requesting access token from " + config.tokenEndpoint, e);
-                        response.sendError(HttpServletResponse.SC_FORBIDDEN, e.getMessage());
-                        return Authentication.SEND_FAILURE;
                     }
+                } catch (OAuthProblemException e) {
+                    if (!mandatory)
+                        return Authentication.UNAUTHENTICATED;
+
+                    log.error("Error received from token endpoint: {} - {}", e.getError(), e.getDescription());
+                    response.sendError(HttpServletResponse.SC_FORBIDDEN);
+                    return Authentication.SEND_FAILURE;
+                } catch (OAuthSystemException e) {
+                    log.error("Error while requesting access token from " + config.tokenEndpoint, e);
+                    response.sendError(HttpServletResponse.SC_FORBIDDEN, e.getMessage());
+                    return Authentication.SEND_FAILURE;
                 }
             }
 
             // if token was obtained, get user info
-            if (clientCredentialsToken != null || accessToken != null) {
+            if ((accessToken != null) || (userInfo != null)) {
                 try {
                     String userId = null;
-                    JsonElement userInfo;
 
-                    if (clientCredentialsToken == null) {
+                    if (userInfo == null) {
```

Comment:
If you make the change I suggested above, need to remove these two tests on `userInfo` here.

Reply:
Refactored code to reduce the number of checks and remove unnecessary checks on userInfo